compiler/src/dotty/tools/dotc/transform/ShortcutImplicits.scala

```diff
@@ -0,0 +1,143 @@
+package dotty.tools.dotc
+package transform
+
+import TreeTransforms._
+import core.DenotTransformers.IdentityDenotTransformer
+import core.Symbols._
+import core.Contexts._
+import core.Types._
+import core.Flags._
+import core.Decorators._
+import core.StdNames.nme
+import core.Names._
+import core.NameOps._
+import ast.Trees._
+import ast.tpd
+import collection.mutable
+
+/** This phase optimizes code using implicit function types, by applying two rewrite rules.
+ *  Let IF be the implicit function type
+ *
+ *      implicit Us => R
+ *
+ *  (1) A method definition
+ *
+ *      def m(xs: Ts): IF = implicit (ys: Us) => E
+ *
+ *  is expanded to two methods:
+ *
+ *      def m(xs: Ts): IF = implicit (ys: Us) => m$direct(xs)(ys)
+ *      def m$direct(xs: Ts)(ys: Us): R = E
```

Comment:
The main reason is that the rewrite is based on the *type* of the result. For implicit function types we have a perfect match. The typing rules guarantee that every function that has an implicit function result type **must** be implemented by a closure. For normal functions this is not guaranteed, so we do not know whether the optimization is globally effective or not. It could make things worse, actually.